refactor: unleash implementation

mc-jones opened the pull request:
The type of this PR is: **Refactor**

<!-- Build / Chore / CI / Docs / Feat / Fix / Perf / Refactor / Revert / Style / Test -->

<!-- If applicable, write the Jira ticket number in square brackets e.g. `[PROJECT-XX]`
     The Jira integration will turn it into a clickable link for you. -->

This PR solves [PHIRE-1620]

### Description

<!-- Implementation description -->

This PR refactors the existing unleash implementation with the following goals: 

- Simplify the code
- Use the `unleash-proxy-client` to retrieve flags and variants when in a browser context. This replaces the existing approach of eagerly loading all flags in `express` `middleware` on every request and using `sharify` to expose them to both server and client.

### Motivations

- Eagerly loading flags and variants breaks metric collection, making observability and maintenance difficult. For example, we are unable to use the unleash admin UI to identify stale flags. 
- The code was unnecessarily complicated. For example, it wrapped the `unleash` `node` SDK in a custom class, introducing an unnecessary level of abstraction. The idea of making it easier to swap in different flag provider easily, while well intentioned, has proven not worth the tradeoff. 
- Using the canonical implementation allows us to rely on `unleash`'s documentation, removing some additional maintenance. 


### Considerations
- This is currently configured to use the [frontend api](https://docs.getunleash.io/reference/front-end-api), which _may_ be sufficient for our use case. The api is the exact same as [unleash edge](https://docs.getunleash.io/reference/unleash-edge), so if we spin up an `edge` instance, we just need to change the `UNLEASH_URL` to point to the instance. I'll need to investigate this a bit more. 

### Usage

When working client side (in `react`), developers can follow the unleash "[how to use](https://docs.getunleash.io/reference/sdks/react)" instructions:
```jsx
import { useFlag, useVariant } from '@unleash/proxy-client-react';

// Check toggle
const TestComponent = () => {
  const enabled = useFlag('demo-feature');

  if (enabled) {
    return <SomeComponent />;
  }
  return <AnotherComponent />;
};

// Check variant
const TestComponent = () => {
  const variant = useVariant('demo-variant');

  if (variant.enabled && variant.name === 'SomeComponent') {
    return <SomeComponent />;
  } else if (variant.enabled && variant.name === 'AnotherComponent') {
    return <AnotherComponent />;
  }
  return <DefaultComponent />;
};
```


When working server side (in `node`), developers just need to retrieve the client first using the `getOrInitUnleashServerClient` function, and then follow the "[how to use](https://docs.getunleash.io/reference/sdks/node#3-check-features)" instructions:
```js
import { getOrInitUnleashServerClient } from "Server/featureFlags/unleashHelpers"

// Get existing unleash client
const unleash = await getOrInitUnleashServerClient()

// Setup context
// NOTE: developers will need to handle passing in the context
// when retrieving flags/variant while working server side. This 
// was done for them in the previous implementation. 
const context = {
  userId: res.locals.sd.CURRENT_USER,
  sessionId: res.locals.sd.SESSION_ID,
};

// Check toggle
if (unleash.isEnabled('demo', context)) {
  console.log('Toggle enabled');
} else {
  console.log('Toggle disabled');
}

// Check variant
const variant = unleash.getVariant('demo-variant', context);

if (variant.name === 'experiment') {
  // do something with the experiment variant...
}
```
### TODO
- [x] Push up modified `env`
- [x] Spike on using `unleash edge` (non-blocking)

I haven't pushed these yet because I wanted the implementation changes to be easier to review:
- [x] Migrate existing flags and experiments
- [x] Refactor test code

[PHIRE-1620]: https://artsyproduct.atlassian.net/browse/PHIRE-1620?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mc-jones commented on src/Server/config.ts:
```diff
@@ -115,7 +114,7 @@ export const TOOLS_URL: any = "https://tools.artsy.net"
 export const TRACK_PAGELOAD_PATHS: any = null
 export const UNLEASH_API: any = null
 export const UNLEASH_APP_NAME: any = null
-export const UNLEASH_INSTANCE_ID: any = null
```
Wasn't used

mc-jones commented on package.json:
```diff
@@ -93,6 +93,7 @@
     "@styled-system/theme-get": "5.1.2",
     "@swc-node/register": "^1.10.9",
     "@swc/helpers": "^0.5.15",
+    "@unleash/proxy-client-react": "^4.5.2",
```
In case it saves others some confusion, I want to clarify a non-obvious point. While the [unleash proxy](https://docs.getunleash.io/reference/unleash-proxy) is deprecated, the `proxy-client`s are not. They are still necessary to interface with our self-hosted `unleash` instances. So we can now spin down the `artsy-unleash-web-proxy` [deployment](https://github.com/artsy/artsy-unleash/blob/main/hokusai/production.yml#L165) and replace it with `unleash edge` or just point request directly to our [web deployment](https://github.com/artsy/artsy-unleash/blob/main/hokusai/production.yml#L4)

joeyAghion replied:
> replace it with unleash edge or just point requests directly to our [web deployment](https://github.com/artsy/artsy-unleash/blob/main/hokusai/production.yml#L4)

I'm confused about this. I thought Edge replaced the proxy. Are you saying Edge is unnecessary and the non-proxy service can also handle requests from clients?

mc-jones replied:
> Are you saying Edge is unnecessary and the non-proxy service can also handle requests from clients?

@joeyAghion, yeah, thats correct. They have a bit of a breakdown [here](https://docs.getunleash.io/reference/front-end-api). The main difference seems to be scalability. If standing up an edge instance up proves straight forward (and I think it will), it may make sense to just do it as part of this work, but for testing and getting agreement, I just started with the frontend api.

mc-jones replied:
Also, given we were hitting it for every request previously, I would think it should be handle the lower throughput.

joeyAghion replied:
Was that actually the case? My understanding was that calls to the client depended only on the already-loaded toggle state, and that state would be refreshed from the server asynchronously based on the configured `refreshInterval`.

mc-jones commented on package.json:
```diff
@@ -178,7 +179,8 @@
     "underscore.string": "3.3.5",
-    "unleash-client": "^3.12.0",
+    "unleash-client": "^6.5.0",
+    "unleash-proxy-client": "^3.7.3",
```
Both `unleash-proxy-client` and `@unleash/proxy-client-react` are needed.

mc-jones commented on src/Server/featureFlags/unleashHelpers.ts:
```diff
@@ -0,0 +1,21 @@
+} from "Server/config"
+import { Unleash, startUnleash } from "unleash-client"
+
+export function getOrInitUnleashServerClient(): Promise<Unleash> {
+  const config = {
+    url: UNLEASH_API,
+    appName: UNLEASH_APP_NAME,
+    refreshInterval: 10000,
+    environment: NODE_ENV,
+    customHeaders: {
+      Authorization: UNLEASH_SERVER_KEY,
+    },
+  }
+
+  return startUnleash(config)
```
This will only ever return a single instance of the `Unleash` client since we're using `startUnleash`, [which implements a singleton](https://github.com/Unleash/unleash-client-node/blob/main/src/index.ts#L16-L30). I'll monitor this in staging, though.

dblandin replied:
It was a bit unclear in the example given in the PR description but if this function returns a promise (as startUnleash does) – I imagine callers will need to wait on that result and handle any errors thrown, as documented [here](https://docs.getunleash.io/reference/sdks/node#synchronous-initialization).

```ts
import { getOrInitUnleashServerClient } from "Server/featureFlags/unleashHelpers"

// Get (or initialize) Unleash client
try {
  const unleash = await getOrInitUnleashServerClient()
} catch (e) {
  // handle error
}
  
```

mc-jones commented on src/System/Boot.tsx:
```diff
@@ -63,6 +65,18 @@ export const Boot: React.FC<
     ...context,
   }
 
+  const unleashConfig = {
+    url: `${sd.UNLEASH_API}frontend/`,
+    clientKey: sd.UNLEASH_CLIENT_KEY,
+    refreshInterval: 15,
+    appName: sd.UNLEASH_APP_NAME,
+    environment: sd.NODE_ENV,
+    context: {
+      userId: props.user?.id,
+      sessionId: sd.SESSION_ID,
+    },
+  }
+
```
Considered moving this into its own file, but prefer it here. If you have strong opinions, LMK.

damassi replied:
A common pattern is to do:

```tsx
import { FlagProvider } from '...'

const FeatureFlagProvider: React.FC<React.PropsWithChildren> = ({ children }) => {
  const { user } = useSystemContext() 
  
  const unleashConfig = {
    url: `${getENV("UNLEASH_API")}frontend/`,
    clientKey: getENV("UNLEASH_CLIENT_KEY"),
    refreshInterval: 15,
    appName: getENV("UNLEASH_APP_NAME"),
    environment: getENV("NODE_ENV"),
    context: {
      userId: user?.id,
      sessionId: getENV("SESSION_ID"),
    },
  }
  
  return <FlagProvider ... >{children}></FlagProvider> 
}

mc-jones replied:
Nice! Will make this change.

mc-jones commented on src/System/Hooks/useFeatureFlag.tsx:
```diff
@@ -1,79 +1,14 @@
-  return variant
-}
-
 export function useTrackFeatureVariant({
```
Variant _tracking_ will remain unchanged.

mc-jones commented on src/Server/bootstrapSharify.ts:
```diff
@@ -84,6 +83,7 @@ export const bootstrapSharify = () => {
       "SEGMENT_AMP_WRITE_KEY",
       "SEGMENT_WRITE_KEY",
       "SENTRY_PUBLIC_DSN",
+      "SESSION_ID",
```
I was surprised this wasn't already in here. I know we have access to it after we the app boots (in this use case its needed before boot), but I couldn't find exactly where that happens.

dblandin replied:
It looks like `SESSION_ID` is added in this [bootstrapSharifyAndContextLocalsMiddleware middleware](https://github.com/artsy/force/blob/7b4ca6cc8a004323022549c39ec25cd87829c5a9/src/Server/middleware/bootstrapSharifyAndContextLocalsMiddleware.ts#L23-L30), executed upon each request whereas `bootstrapSharify` is [only executed once](https://github.com/artsy/force/blob/7b4ca6cc8a004323022549c39ec25cd87829c5a9/src/middleware.ts#L57) upon server initialization so I don't think this addition will have any effect as we never seed `SESSION_ID` from the process environment.

damassi reviewed (dismissed):
This is such a nice simplifying PR 👍

joeyAghion commented on src/Server/config.ts:
```diff
@@ -115,7 +114,7 @@ export const TOOLS_URL: any = "https://tools.artsy.net"
 export const UNLEASH_API: any = null
 export const UNLEASH_APP_NAME: any = null
-export const UNLEASH_INSTANCE_ID: any = null
+export const UNLEASH_CLIENT_KEY: any = null
```
Just flagging this new configuration that will be required pre-release.

chr-tatu reviewed (dismissed):
So much simpler and shorter implementation. Thanks for taking the initiative to do this!

dblandin commented on src/System/Boot.tsx:
```diff
@@ -63,6 +65,18 @@ export const Boot: React.FC<
+  const unleashConfig = {
+    url: `${sd.UNLEASH_API}frontend/`,
+    clientKey: sd.UNLEASH_CLIENT_KEY,
+    refreshInterval: 15,
```
A bit confusing that this value is taken in seconds whereas the node library config value with the same name takes milliseconds. Any reason to depart from the library default of 30s as the refreshInterval?

mc-jones replied:
Good call. Have added in [90f2431](https://github.com/artsy/force/pull/15263/commits/90f243108dd075e48bc9ba0a0b4c41c1abafda6d) and set to `30` seconds. No specific reason for `15` other than following documented initialization [example](https://docs.getunleash.io/reference/sdks/react#initialize-the-client).

dblandin commented on src/System/Boot.tsx:
```diff
@@ -63,6 +65,18 @@ export const Boot: React.FC<
+    clientKey: sd.UNLEASH_CLIENT_KEY,
+    refreshInterval: 15,
+    appName: sd.UNLEASH_APP_NAME,
+    environment: sd.NODE_ENV,
```
I think the only environments we support on the Unleash side are `development` and `production`. Should we coerce this value into one of those supported values?

We could infer from NODE_ENV but it be clearer and more flexible to use an explicit `UNLEASH_ENVIRONMENT` config value which we can default to `development` unless targeting a production environment.

mc-jones replied:
Agree! good call out. Was conflating runtime environment with how we've configured `unleash`'s environment. Changed [229a483](https://github.com/artsy/force/pull/15263/commits/229a483081f47867c472e9492fabf9e04dfc0562)

dblandin commented on src/Typings/sharify.d.ts:
```diff
@@ -101,6 +100,10 @@ declare module "sharify" {
       THIRD_PARTIES_DISABLED: boolean
       TOOLS_URL: string
       TRACK_PAGELOAD_PATHS: string
+      UNLEASH_API: string
```
small nit but maybe `UNLEASH_API_URL` if this is also meant to contain a URL value?

mc-jones replied:
Good call. Agree. Changed in [eb65921](https://github.com/artsy/force/pull/15263/commits/eb65921130a2bb6d59d38e0d8192594898aee451)

dblandin reviewed (commented):
Awesome work here 💯

mc-jones commented on src/System/Contexts/FeatureFlagContext.tsx:
```diff
@@ -0,0 +1,42 @@
+
+  let unleashConfig: IConfig
+
+  if (process.env.NODE_ENV === "test") {
+    unleashConfig = {
+      url: "http://mock-unleash-url",
+      clientKey: "mock-client-key", // pragma: allowlist secret
+      refreshInterval: 0,
+      appName: "mock-app-name",
+      context: {
+        userId: "mock-user-id",
+        sessionId: "mock-session-id",
+      },
+    }
```
This feels... hacky and I feel like I'm overlooking something obvious. If we have a better pattern for config required during test runs, LMK. 

cc: @damassi

mc-jones replied:
Also, I'm blocked on this final [smoke-test error](https://app.circleci.com/pipelines/github/artsy/force/68908/workflows/a2e22c01-3930-4a27-a518-32a971264e87/jobs/491939).  Running locally,  I can see that we're throwing the following error server side, while trying to call `@unleash/proxy-client-react...` which makes sense given its a hook and we're outside the client context. But its unclear why were trying to render `Users/mcj/code/force/src/Components/ViewInRoom/ViewInRoom.tsx`. I'll keep debugging, but if anyone has a quick answer, lmk. 

```
Server/middleware/errorHandlerMiddleware | TypeError: Cannot destructure property 'isEnabled' of 'o.useContext(...)' as it is null.
    at M (/Users/mcj/code/force/node_modules/@unleash/proxy-client-react/src/useFlag.ts:5:11)
    at /Users/mcj/code/force/src/Components/ViewInRoom/ViewInRoom.tsx:79:1
    at renderWithHooks (/Users/mcj/code/force/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:5662:16)
    at renderForwardRef (/Users/mcj/code/force/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:5857:18)
    at renderElement (/Users/mcj/code/force/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6020:11)
    at renderNodeDestructiveImpl (/Users/mcj/code/force/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6119:11)
    at renderNodeDestructive (/Users/mcj/code/force/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6091:14)
    at renderNode (/Users/mcj/code/force/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6274:12)
    at renderChildrenArray (/Users/mcj/code/force/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6226:7)
    at renderNodeDestructiveImpl (/Users/mcj/code/force/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:6156:7) 
```

damassi replied:
@mc-jones - sorry just now seeing this! 

Whats wrong with standard jest mocks? Something like the below, placed in tests.tsx...

```
jest.mock("System/Contexts/FeatureFlagContext", () => ({
  FeatureFlagProvider: ({ children }) => children,
  useOtherThingThatMightBeSomewhere: jest.fn() 
}))
```

And then when _explicitly_ testing the file, can unmock:

jest.unmock("...") 

So yah, that should allow one to skip this branching.

mc-jones replied:
@damassi 🙏 - I was misunderstanding how our globals were working, somehow missing the `tests.tsx` file. Fixed in [59a4cba](https://github.com/artsy/force/pull/15263/commits/59a4cba293982cdbb2108e3100a5e2833b42b3b9)

mc-jones commented:
Ok, thanks for the feedback all. This is now ready for final review. I've incorporated the suggested changes, and modified existing flags and tests to work with the new implementation. 

Regarding the[ frontend api vs unleash edge question](https://github.com/artsy/force/pull/15263#discussion_r1974462393), I'm planning to start with the frontend api. I believe that our  instance should be able to handle the traffic, as we are hitting that instance directly, right now, on every force request. Now that we're switching to only requesting the flag when needed, I expect the overall volume of request to the instance to go down. 

If this looks good to everyone, I'll communicate the changes to developers and start the process of rolling this out.

damassi commented on src/Server/__tests__/featureFlagProvider.jest.ts:
```diff
@@ -1,42 +0,0 @@
-  createFeatureFlagService,
-  registerFeatureFlagService,
-} from "Server/featureFlags/featureFlagService"
-import type { Variant } from "unleash-client"
```
✌️

damassi commented on src/middleware.ts:
```diff
@@ -170,10 +164,6 @@ export function initializeMiddleware(app) {
 
-  // Need sharify for unleash
-  registerFeatureFlagService(UnleashService, UnleashFeatureFlagService)
-  app.use(featureFlagMiddleware(UnleashService))
```
✌️

damassi reviewed (approved):
Everything looks simple and ideal now, and so much code deleted! 💯

The only possible issue I could see happening in the future (we should cross this bridge when it comes up, however, and not for this PR as we're not actually using this anywhere) would be at the router level, because now we no longer have access to a unified featureFlag mechanism from within the router context (both on client and server).

For example, this is how it once worked:

```tsx
const routes = [
  {
    path: "/foo",
    query: graphql`
      const myQuery($enableFoo: Boolean!) {
        artwork(id: "foo") {
          ...SomeFragment_artwork @include(if: $enableFoo)
        }
      }
    `,
    prepareVariables: (params, { context }) => {
      return {
        skipFoo: context.featureFlags.skipFoo
      }
    }
  }
]
```

Just a heads up, thinking experiments and the like. Something like this would need to be SSR safe and available at this level if we want things to work smoothly. 

Otherwise 🙆‍♂️

joeyAghion commented:
> given we were hitting it for every request previously, I would think it should be handle the lower throughput

Was this actually the case? My understanding was that calls to the client depended only on the already-loaded toggle state, and that state would be refreshed from the server asynchronously based on the configured `refreshInterval`.
